src/models/shelf/spec/encoding.ts

```diff
@@ -65,10 +65,10 @@ export interface ShelfAnyEncodingDef extends ShelfFieldDef {
   channel: SHORT_WILDCARD;
 }
 
+export type ShelfValueDef = ValueDef;
+
 export type SpecificEncoding = {
-  // TODO: ShelfFieldDef | ValueDef
-  // (Just use ValueDef, no need for special ShelfValueDef)
-  [P in Channel]?: ShelfFieldDef;
+  [P in FieldDefOnlyChannels]?: ShelfFieldDef
```

Comment:
I don't recall having `FieldDefOnlyChannels`. 

You can use `PositionScaleChannel`, `NonPositionScaleChannel`.
You can then manually add `type NonPositionNonScaleChannel = 'text' | 'tooltip' | 'detail' | 'order'`.